sound/soc/sof/intel/hda.c

```diff
@@ -53,67 +53,72 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
 	struct snd_soc_dai *d = params_data->dai;
-	struct sof_ipc_dai_config config;
-	struct sof_ipc_reply reply;
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
 	int link_id = params_data->link_id;
 	int alh_stream_id = params_data->alh_stream_id;
-	int ret;
-	u32 size = sizeof(config);
-
-	memset(&config, 0, size);
-	config.hdr.size = size;
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.type = SOF_DAI_INTEL_ALH;
-	config.dai_index = (link_id << 8) | (d->id);
-	config.alh.stream_id = alh_stream_id;
-
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config.hdr.cmd, &config, size, &reply,
-				 sizeof(reply));
-	if (ret < 0) {
-		dev_err(sdev->dev,
-			"error: failed to set DAI hw_params for link %d dai->id %d ALH %d\n",
-			link_id, d->id, alh_stream_id);
+
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
+		if (!sof_dai->cpu_dai_name)
+			continue;
+
+		config = sof_dai->dai_config;
+		if (!config) {
+			dev_err(dev, "error: no config for DAI %s\n",	sof_dai->name);
+			return -EINVAL;
+		}
+
+		if (!strcmp(d->name, sof_dai->cpu_dai_name) &&
+		    substream->stream == sof_dai->comp_dai.direction) {
+			config->dai_index = (link_id << 8) | (d->id);
+			config->alh.stream_id = alh_stream_id;
+			return 0;
+		}
 	}
 
-	return ret;
+	dev_err(dev, "error: failed to update DAI config for: %s\n", d->name);
+	return -EINVAL;
 }
 
+/* update the DAI config with the invalid stream_id. The IPC will be sent during FE hw_params */
 static int sdw_free_stream(struct device *dev,
 			   struct sdw_intel_stream_free_data *free_data)
 {
+	struct snd_pcm_substream *substream = free_data->substream;
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
 	struct snd_soc_dai *d = free_data->dai;
-	struct sof_ipc_dai_config config;
-	struct sof_ipc_reply reply;
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
 	int link_id = free_data->link_id;
-	int ret;
-	u32 size = sizeof(config);
-
-	memset(&config, 0, size);
-	config.hdr.size = size;
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.type = SOF_DAI_INTEL_ALH;
-	config.dai_index = (link_id << 8) | d->id;
-	config.alh.stream_id = 0xFFFF; /* invalid value on purpose */
-
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config.hdr.cmd, &config, size, &reply,
-				 sizeof(reply));
-	if (ret < 0) {
-		dev_err(sdev->dev,
-			"error: failed to free stream for link %d dai->id %d\n",
-			link_id, d->id);
+	int alh_stream_id = 0xFFFF; /* invalid value on purpose */
+
+	list_for_each_entry(sof_dai, &sdev->dai_list, list) {
+		if (!sof_dai->cpu_dai_name)
+			continue;
+
+		config = sof_dai->dai_config;
+		if (!config) {
+			dev_err(dev, "error: no config for DAI %s\n",	sof_dai->name);
+			return -EINVAL;
+		}
+
+		if (!strcmp(d->name, sof_dai->cpu_dai_name) &&
+		    substream->stream == sof_dai->comp_dai.direction) {
+			config->dai_index = (link_id << 8) | (d->id);
```

Comment:
we don't need to update dai_index again for free function just reset stream_id to invalid one. @plbossart @bardliao ?

Reply:
Right, IIRC we set it to 0xffffffff before.

Reply:
we do need to set the dai_index otherwise the IPC will fail.

Reply:
@RanderWang Did you mean we don't need to set dai_index since it is not changed?

Reply:
sure,  I think this dai index should not be changed